### PR TITLE
Add Elliptic Curve variants for JWS and JWK utilities

### PR DESCRIPTION
**Pull request by jarrodmoldrich**
Hello Airside/Mohemian team,

I needed EC algorithms for signing/verifying JWTs, and for decoding/encoding JWKs, so I thought it prudent to fork JOSESwift.  The functionality for JWS and JWK should be complete and working, but I've yet to mirror your comprehensive test suite for RSA to the EC domain.  I will be doing so over the coming weeks 🙂 

Anecdotally, JWK decoding/encoding and JWS verification are working for `ES256` public keys for the project I'm working on.

This is a WIP, tell me if there's any issues with it.  It would be good to see this eventually hit upstream so we can retarget to master.

Cheers,

Jarrod

### CODE REVIEW

**Comment by daniel-mohemian** [comment]
👋 Hi Jarrod

That's great news! 🎉 

We were actually planning on supporting EC keys in our JWK implementation and consequently using them for signing and/or encryption for the mid-term future, so any help in that domain is greatly appreciated.

@carol-mohemian and I will have a look at your code over the next days and provide some feedback. At a first glance, it already looks really great!

Once we have some progress on the test suite, we'd be happy to merge this to master and put it in the next release of JOSESwift. 😎

Thanks again, you're awesome
- Daniel

**Comment by jarrodmoldrich** [comment]
Hi @daniel-mohemian 👋 

Thank you for the warm welcome.  I'll be pecking away at this in my downtime, so it might be a 4-8 week turn-around for the full test suite.  Let me know if that conflicts with you, and I'll see what I can do.

Cheers,

Jarrod

File: JOSESwift/Sources/ECVerifier.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//  ---------------------------------------------------------------------------
+//
+
+import Foundation
+
+/// A `Verifier` to verify a signature created with a elliptic curve algorithm.
```
Comment (carol-mohemian) [inline JOSESwift/Sources/ECVerifier.swift]: `an elliptic curve..`

File: JOSESwift/Sources/Signer.swift [hunk 2]
```diff
@@ -51,6 +52,8 @@ public struct Signer<KeyType> {
             }
             // swiftlint:disable:next force_cast
             self.signer = RSASigner(algorithm: signingAlgorithm, privateKey: privateKey as! RSASigner.KeyType)
+        default:
```
Comment (carol-mohemian) [inline JOSESwift/Sources/Signer.swift]: Why you don't add the ECSigner here?

File: JOSESwift/Sources/Verifier.swift [hunk 3]
```diff
@@ -53,6 +54,12 @@ public struct Verifier<KeyType> {
             }
             // swiftlint:disable:next force_cast
             self.verifier = RSAVerifier(algorithm: verifyingAlgorithm, publicKey: publicKey as! RSAVerifier.KeyType)
+        case .ES256,.ES384,.ES521:
+            guard type(of: publicKey) is ECVerifier.KeyType.Type else {
+                return nil
+            }
+            self.verifier = ECVerifier(algorithm: verifyingAlgorithm, publicKey: publicKey as! ECVerifier.KeyType)
+
```
Comment (carol-mohemian) [inline JOSESwift/Sources/Verifier.swift]: Empty new line.

**Comment by daniel-mohemian** [comment]
Sure, sounds great @jarrodmoldrich. 👍 

If you want, we can also help out with some tests. It would be cool if you could create a basic test suite and maybe use some examples from the relevant RFCs as the test data. We can then still expand the test suite after merging this PR.

As said, the implementation looks really great already. 🎉 
It would be cool if you could work in the comments that @carol-mohemian had if you agree. Please also make sure to put the license into every newly generated file and make sure that it says `Copyright 2018 Airside Mobile Inc.`.

Let me know if you need anything!

File: JOSESwift/Sources/Algorithms.swift [hunk 4]
```diff
@@ -26,9 +27,13 @@ import Foundation
 /// An algorithm for signing and verifying.
 ///
 /// - RS512: [RSASSA-PKCS1-v1_5 using SHA-512](https://tools.ietf.org/html/rfc7518#section-3.3)
+/// - ES512: [ECDSA P-512 using SHA-512](https://tools.ietf.org/html/rfc7518#section-3.4)
 public enum SignatureAlgorithm: String {
     case RS256 = "RS256"
     case RS512 = "RS512"
+    case ES256 = "ES256"
+    case ES384 = "ES384"
+    case ES521 = "ES521"
```
[the remark this answers is not in the document]
Comment (jarrodmoldrich) [reply JOSESwift/Sources/Algorithms.swift]: Ah yes, got my `P-521`s and `ES512`s mixed up 🤦‍♂️

File: JOSESwift/Sources/JOSESwiftError.swift [hunk 5]
```diff
@@ -21,7 +22,17 @@ public enum JOSESwiftError: Error {
...
     case modulusNotBase64URLUIntEncoded
     case exponentNotBase64URLUIntEncoded
     case privateExponentNotBase64URLUIntEncoded
+
+    // EC coding errors
+    case xNotBase64URLUIntEncoded
+    case yNotBase64URLUIntEncoded
+    case privateKeyNotBase64URLUIntEncoded
+    case invalidCurveType
+    case compressedCurvePointsUnsuported
```
[the remark this answers is not in the document]
Comment (jarrodmoldrich) [reply JOSESwift/Sources/JOSESwiftError.swift]: Good catch, I'll fix the typo.

We're specifically not supporting compressed curve points at all, so I'm not sure the other two suggestions effectively convey this limitation.

[the remark this answers is not in the document]
Comment (jarrodmoldrich) [reply JOSESwift/Sources/JOSESwiftError.swift]: I see what you mean about the "error below".  GitHub doesn't provide surrounding context, but I'd agree that the second error could be `invalidCurvePointOctetLength`.

**Comment by jarrodmoldrich** [comment]
@daniel-mohemian It probably makes sense that I create a minimal test suite.  I think the sooner this PR gets merged, the better, as there's a lot of opportunity to refactor out much of the shared functionality between the key types.

The Apache license seemed to me a little ambiguous about the copyright notice, but on a second read, a PR classes more as a 'Contribution' than a 'Derivative Work' and so I've changed to your copyright notice 

I'll be AFK for the next couple of weeks, so I probably won't start on the test suite until August.  Feel free to leave feedback in the meantime 🙂

File: JOSESwift/Sources/CryptoImplementation/EC.swift [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        case 521:
+            return .P521
+        default:
+            return nil
+        }
+    }
+
+    static func fromCoordinateOctetLength(_ length: Int) -> ECCurveType? {
+        switch length {
+        case 32:
```
[the remark this answers is not in the document]
Comment (carol-mohemian) [reply JOSESwift/Sources/CryptoImplementation/EC.swift]: I just thought that you can use the `ECCurveType` enum and it's values for `keyBitLength/coordinateOctetLength` to not hardcode those length values again.

**Comment by daniel-mohemian** [comment]
Hi @jarrodmoldrich 👋 

I just wanted to check in with you to see if you need any assistance with the tests. If you want we can also take over a part of the test suite. If you're back online already, let me know what you think, otherwise, enjoy your time off. 🙂

**Comment by daniel-mohemian** [comment]
That sounds great. Just let me know if you need anything. We're not in a hurry, end of September sounds like a plan! 😄

**Comment by jarrodmoldrich** [comment]
@daniel-mohemian It's a good thing you're not in a hurry!  Things were a bit hectic for me this past month, but I should finally have some free time at the end of this coming week.  I was going to make a start by adding test data to `CryptoTestCase.swift`, maybe separating/refactoring any general functionality if it makes sense, and then building the related test cases - any input is appreciated.  So hopefully there should be some movement on this PR soon.

**Comment by daniel-mohemian** [comment]
That's fine, @jarrodmoldrich.

Refactoring and expanding the base class for crypto tests - `CryptoTestCase.swift` - sound great! Don't be afraid to rearrange things if you see the need. The tests are still a bit chaotic here and there. As you said, you can then add some EC specific tests.

- `Crypto/` contains tests for the crypto primitives. It would be a good place to test your `EC`, `ECSigner`, `ECVerifier` there.
- `JWS/` contains JWS specific tests such as roundtrips and serialization. Maybe try to add some tests that serialize, deserialize, sign and verify a JWS with EC keys.
- `JWK/` contains tests specific to the JWK representations of keys. It would be great if you could test the conversion from `SecKey` <->`Data` <-> `ECPublic/PrivateKey` there.

Let me know if you need anything!

**Comment by daniel-mohemian** [comment]
I see you added quite some tests @jarrodmoldrich. 🎉

Please ping me if they are ready to be looked at.

PS: Don't worry about the merge conflicts, I can take care of that for you before merging.

**Comment by jarrodmoldrich** [comment]
Sorry for the delay, @daniel-mohemian .  I've been occupied the last few months, but I'll have this ready for your review at my self-imposed deadline of Jan 14.

**Comment by jarrodmoldrich** [comment]
@daniel-mohemian Is `SJCommonCrypto` gone for good?  The EC signing implementation is using CommonCrypto to create a digest.  I could fish around to see if there's some new way to write the signing algorithm, but it would be much easier to drag CC back in.

**Comment by carol-mohemian** [comment]
Hey @jarrodmoldrich 👋 Please see the PR description of #126.

TL;DR
> Swift 4.2 introduced a native common crypto module which we can/must use now.

**Comment by jarrodmoldrich** [comment]
Thanks @carol-mohemian .  I think it's finally ready for review.

**Comment by daniel-mohemian** [comment]
Awesome @jarrodmoldrich, I hope we can provide feedback soon! We're also quite packed with work but I'll do my best to get the pull request merged over the next weeks. At the very latest in February I'll have some more time on my hands. I hope this is ok for you.

We already looked over it when you first opened the pull request and it looked really good back then. 👌 

Oh and happy new year to you too! 🎉

File: JOSESwift/Sources/JWKSetCodable.swift [hunk 7]
```diff
@@ -27,6 +28,43 @@ enum JWKSetParameter: String, CodingKey {
...
+
+internal enum JWKTypeError: Error {
+    case typeIsECPrivate
+    case typeIsECPublic
+    case typeIsRSAPublic
+    case typeIsRSAPrivate
+    case typeIsUnknown
+}
+
+class JWKBase: Decodable {
```
[the remark this answers is not in the document]
Comment (jarrodmoldrich) [reply JOSESwift/Sources/JWKSetCodable.swift]: Looking back at an earlier commit I had modified `init(from decoder: Decoder)` to look like it is below, but during the integration I must have reverted it to it's original style.  I think it looks a little neater the way I've done it, so I could either revert the approach once more, or remove `JWKBase`.  I'll defer to your judgement.

```
            do {
                _ = try keyContainer.decode(JWKBase.self)
            } catch JWKTypeError.typeIsECPublic {
                key = try keyContainer.decode(ECPublicKey.self)
            } catch JWKTypeError.typeIsECPrivate {
                key = try keyContainer.decode(ECPrivateKey.self)
            } catch JWKTypeError.typeIsRSAPublic {
                key = try keyContainer.decode(RSAPublicKey.self)
            } catch JWKTypeError.typeIsRSAPrivate {
                key = try keyContainer.decode(RSAPrivateKey.self)
            }
```

[the remark this answers is not in the document]
Comment (daniel-mohemian) [reply JOSESwift/Sources/JWKSetCodable.swift]: I like your approach, let's go for that! 

One suggestion from me would be to use the raw values of `JWKKeyType` case statements in `JWKBase` like this:

``` swift
switch key {
case JWKKeyType.RSA.rawValue:
    if d == nil {
        throw JWKTypeError.typeIsRSAPublic
    } else {
        throw JWKTypeError.typeIsRSAPrivate
    }
}
```

What do you think?

File: JOSESwift/Sources/JWSHeader.swift [hunk 8]
```diff
@@ -64,13 +64,16 @@ public struct JWSHeader: JOSEHeader {
...
 
-        // Forcing the try is ok here, since "alg" is the only required header parameter.
+        // Forcing the try is ok here, since [String: String] can be converted to JSON and "alg" is the only required
+        // header parameter, which should pass the guard conditions in the main initializer
         // swiftlint:disable:next force_try
-        try! self.init(parameters: parameters, headerData: headerData)
+        try! self.init(parameters: parameters)
+    }
+
+    public init(parameters: [String: Any]) throws {
```
[the remark this answers is not in the document]
Comment (daniel-mohemian) [reply JOSESwift/Sources/JWSHeader.swift]: That's cool. 😎 

We added support for optional parameters recently via #110. That must have been after you forked. So you can now either set at initializer level via your delegating initializer or afterwards via setters. 👍 

I'll add a similar initializer to the `JWEHeader` for consistency.

File: Tests/ECSignerTests.swift [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    private func _testSigning(algorithm: SignatureAlgorithm, keyData: ECTestKeyData) {
+        let messageData = message.data(using: .utf8)!
+        let signer = ECSigner(algorithm: algorithm, privateKey: keyData.privateKey)
+        let signature = try! signer.sign(messageData)
+        let signature2 = try! signer.sign(messageData)
+
+        // As any two signing invocations will have different nonces, it is impossible to use pre-generated data from a
+        // trusted implementation (e.g. openssl) to verify the signature.  Instead we will validate by verifying the
+        // signature.  This assumes that the verification implementation is correct.
```
Comment (daniel-mohemian) [inline Tests/ECSignerTests.swift]: 👍

File: Tests/JWKECEncodingTests.swift [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                    y: keyData.expectedYCoordinateBase64Url,
+                    privateKey: keyData.expectedPrivateBase64Url,
+                    additionalParameters: ["alg": keyData.signatureAlgorithm, "kid": Consts.kid, "breeze": "through"]
+            )
+
+            let jsonData = try? JSONEncoder().encode(jwk)
+            let dict = checkAndGetDictionary(jsonData: jsonData)
+
+            checkRegularParameters(dict: dict, keyData: keyData)
+            XCTAssertEqual(dict["d"] as? String ?? "", keyData.expectedPrivateBase64Url)
```
Comment (daniel-mohemian) [inline Tests/JWKECEncodingTests.swift]:
``` suggestion
            XCTAssertEqual(dict["d"] as? String ?? "", keyData.expectedPrivateBase64Url)
            XCTAssertNil(dict["breeze"])
```

File: JOSESwift/Sources/JWKSetCodable.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (jarrodmoldrich) [reply JOSESwift/Sources/JWKSetCodable.swift]: Good point, yes, I should be doing that.  Updated, along with a fix for the catch mechanism I just noticed.  (Not using optional `try?`.)

**Comment by jarrodmoldrich** [comment]
@daniel-mohemian Implemented your suggestions and integrated the decoding pattern with a couple of changes above what we discussed, hopefully they are all good 🤞

**Comment by jarrodmoldrich** [comment]
Likewise, @daniel-mohemian  🙂

**Comment by daniel-mohemian** [comment]
Your contribution is now released as part of [1.5.0](https://github.com/airsidemobile/JOSESwift/releases/tag/1.5.0). 🎉
